### PR TITLE
image: boot from dm-verity protected partition

### PR DESCRIPTION
**Pull request by danmihai1**
Mount this partition and use it as root - without the need to use
an initramfs.

Fixes: #7906

### CODE REVIEW

**Review by arronwy — APPROVED** [review]
Thanks @danmihai1 cool, LGTM! since this PR support both kernel init and initramfs, we still can use initramfs for more flexible scenarios like get encrypted key and attestation in initramfs to support encrypted rootfs

**Review by fidencio — APPROVED** [review]
lgtm, conditionally to have the TDX tests passing before merging which was agreed with @danmihai1!

**Comment by fidencio** [comment]
@danmihai1, mind to solve the conflicts so we can have this one merged?

**Comment by danmihai1** [comment]
> @danmihai1, mind to solve the conflicts so we can have this one merged?

@fidencio this is difficult because I believe:

1. We currently have a single image that has MEASURED_ROOTFS enabled (the TDX image), and

2. A few weeks ago, https://github.com/kata-containers/kata-containers/commit/48bdca4c49010d5c347c255b342e0918cc1cb5c7 added a test that makes sure that the TDX image *does not* use the DM_VERITY_FORMAT=kernelinit format added here.


So, unless I am missing something important here, I propose:

3. Change the cbl-mariner tests to use an image file having MEASURED_ROOTFS=yes and DM_VERITY_FORMAT=kernelinit.

4. Find a different way to test the initrd annotation - now that cbl-mariner will not test that annotation anymore.

(3) is important for Microsoft regardless of the blocking new test discussed above, because (3) is the configuration used by AKS. 


Other possibilities:

5. Add MEASURED_ROOTFS=yes and DM_VERITY_FORMAT=kernelinit to some other image. Which image would that be?

6. Move the older-style MEASURED_ROOTFS=yes that k8s-measured-rootfs.bats depends on to some other image. Which image would that be?

Thoughts? Thanks!

**Comment by fidencio** [comment]
We briefly discussed this in the kata-dev channel, and going for 3 is the best approach.

Thanks, @danmihai1!

**Comment by wainersm** [comment]
Added the "do-not-merge" because I'm reviewing this at that exact moment. However I noticed it triggered the CI again, sorry :(

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries-in-docker.sh [hunk 1]
```diff
@@ -119,6 +120,7 @@ docker run \
 	--env TOOLS_CONTAINER_BUILDER="${TOOLS_CONTAINER_BUILDER}" \
 	--env VIRTIOFSD_CONTAINER_BUILDER="${VIRTIOFSD_CONTAINER_BUILDER}" \
 	--env MEASURED_ROOTFS="${MEASURED_ROOTFS}" \
+	--env DM_VERITY_FORMAT="${DM_VERITY_FORMAT:-}" \
```
Comment (wainersm) [inline tools/packaging/kata-deploy/local-build/kata-deploy-binaries-in-docker.sh:123]: Nit: `--env DM_VERITY_FORMAT="${DM_VERITY_FORMAT}" \`

File: tools/packaging/kernel/build-kernel.sh [hunk 2]
```diff
@@ -110,6 +112,7 @@ Options:
 	-t <hypervisor>	: Hypervisor_target.
 	-u <url>	: Kernel URL to be used to download the kernel tarball.
 	-v <version>	: Kernel version to use if kernel path not provided.
+	-V <format>	: dm-verity partition format for measured guest rootfs - kernelinit or veritysetup.
```
Comment (wainersm) [inline tools/packaging/kernel/build-kernel.sh:115]: Nit: mention it defaults to `veritysetup`

**Comment by wainersm** [comment]
@danmihai1 finished my review. I didn't look at the Go changes but rather the build scripts only. Apart from a few nits, everything looks good (at least I could not find any bug -- damn it!:). I got one question for you, though, see below pls.

I built the vanilla image, kernel and shim-v2 enabling measured rootfs with kernelinit:

```
$ make MEASURED_ROOTFS=yes DM_VERITY_FORMAT=kernelinit  rootfs-image-tarball
$ make MEASURED_ROOTFS=yes DM_VERITY_FORMAT=kernelinit kernel-tarball
$ cd src/runtime
$ make clean
$ make MEASURED_ROOTFS=yes DM_VERITY_FORMAT=kernelinit shim-v2-tarball
```

I checked the `dm-mod.create` entry in `kernel_params` in `configuration-qemu.toml` was generated. Good. But caught my attention the `@ROOTFS_DEVICE@ @VERITY_DEVICE@` markers. Should they had been eval to values?

```
[wmoschet@wainer-laptop shim-v2]$ grep ^kernel_params destdir/opt/kata/share/defaults/kata-containers/configuration-qemu.toml
kernel_params = " dm-mod.create=\"dm-verity,,,ro,0 512000 verity 1 @ROOTFS_DEVICE@ @VERITY_DEVICE@ 4096 4096 64000 0 sha256 628f62d1026f653aed96d91bb46b128f8b3f2ca9d7a70ebe26418d25030a62e7 874fddf40af480917d035cf63da2271220fd6604cdd862d73fbab81624a52a29\""
```

Last but not least, I didn't give it a try but I expect that by enabling it for cbl-mariner image/kernel it can be proved on CI it works :)

**Comment by wainersm** [comment]
> We briefly discussed this in the kata-dev channel, and going for 3 is the best approach.
> 
> Thanks, @danmihai1!

Will the tests for 3 be in a separate PR?

**Comment by danmihai1** [comment]
> > We briefly discussed this in the kata-dev channel, and going for 3 is the best approach.
> > Thanks, @danmihai1!
> 
> Will the tests for 3 be in a separate PR?

My plan:
1. Change cbl-mariner from initrd to a vanilla image format, in a separate PR.
2. Enable verity and test here for that future image.

**Comment by danmihai1** [comment]
> @danmihai1 finished my review. I didn't look at the Go changes but rather the build scripts only. Apart from a few nits, everything looks good (at least I could not find any bug -- damn it!:). I got one question for you, though, see below pls.
> 
> I built the vanilla image, kernel and shim-v2 enabling measured rootfs with kernelinit:
> 
> ```
> $ make MEASURED_ROOTFS=yes DM_VERITY_FORMAT=kernelinit  rootfs-image-tarball
> $ make MEASURED_ROOTFS=yes DM_VERITY_FORMAT=kernelinit kernel-tarball
> $ cd src/runtime
> $ make clean
> $ make MEASURED_ROOTFS=yes DM_VERITY_FORMAT=kernelinit shim-v2-tarball
> ```
> 
> I checked the `dm-mod.create` entry in `kernel_params` in `configuration-qemu.toml` was generated. Good. But caught my attention the `@ROOTFS_DEVICE@ @VERITY_DEVICE@` markers. Should they had been eval to values?
> 
> ```
> [wmoschet@wainer-laptop shim-v2]$ grep ^kernel_params destdir/opt/kata/share/defaults/kata-containers/configuration-qemu.toml
> kernel_params = " dm-mod.create=\"dm-verity,,,ro,0 512000 verity 1 @ROOTFS_DEVICE@ @VERITY_DEVICE@ 4096 4096 64000 0 sha256 628f62d1026f653aed96d91bb46b128f8b3f2ca9d7a70ebe26418d25030a62e7 874fddf40af480917d035cf63da2271220fd6604cdd862d73fbab81624a52a29\""
> ```
> 
> Last but not least, I didn't give it a try but I expect that by enabling it for cbl-mariner image/kernel it can be proved on CI it works :)

The values of @ROOTFS_DEVICE@ and its friends are available just at run time - so you need to read the new Go code to find how these work.

File: src/runtime/virtcontainers/hypervisor.go [hunk 3]
```diff
@@ -125,8 +127,20 @@ const (
 	EROFS RootfsType = "erofs"
 )
 
-func GetKernelRootParams(rootfstype string, disableNvdimm bool, dax bool) ([]Param, error) {
+const (
+	// DeviceMapperCreate is a kernel parameter used for creating /dev/dm-0
+	// from two image partitins - one containing the rootfs files and the
+	// other containing the corresponding dm-verity Merkle tree. /dev/dm-0
+	// gets created during kernel devive mapper initialization, after the
```
Comment (wainersm) [inline src/runtime/virtcontainers/hypervisor.go:134]: s/devive/device/

File: src/runtime/virtcontainers/hypervisor.go [hunk 4]
```diff
@@ -125,8 +127,20 @@ const (
 	EROFS RootfsType = "erofs"
 )
 
-func GetKernelRootParams(rootfstype string, disableNvdimm bool, dax bool) ([]Param, error) {
+const (
+	// DeviceMapperCreate is a kernel parameter used for creating /dev/dm-0
+	// from two image partitins - one containing the rootfs files and the
```
Comment (wainersm) [inline src/runtime/virtcontainers/hypervisor.go:132]: s/partitins/partitions/

**Pull request closed without merging**